### ISSUE TITLE
Support disabling initramfs networking via Ignition

### ISSUE THREAD

**Issue by stbenjam** [coreos/ignition#979]
# Feature Request #

Support running part of ignition before networking comes up

## Environment ##

RHCOS

What hardware/cloud provider/hypervisor is being used to run Ignition?

Baremetal

## Desired Feature ##

For baremetal IPI platform in OpenShift, we load ignition from disk.  We would like to include network configuration that happens before NetworkManager starts. For example, we want to disable dhcp on some interfaces. If we can't do that, then on some hosts that could have upwards of 10 (!) interfaces, booting takes an extraordinary long time.

Also, when an ignition contains an 'append' with a network source, it's conceivable we need to do some pre-configuration to make that network endpoint accessible (for example, we need to configure a VLAN on an interface).

## Other Information ##

It's almost like we need a pre-ignition tool (preflight?) analogous to what afterburn is doing post-ignition.

**Comment by stbenjam** [comment]
@cgwalters What do you think of something like this?

**Comment by hardys** [comment]
To link this to a real (OpenShift) use-case, see https://bugzilla.redhat.com/show_bug.cgi?id=1824331 - in some baremetal environments the external network can be on a tagged vlan, so you have to do some configuration via ignition before the `append` (in v2.x) URL can be evaluated.  I guess a similar issue exists with the `merge` feature in v3.

**Comment by cgwalters** [comment]
A lot of overlap here with https://github.com/openshift/enhancements/pull/291

**Comment by cgwalters** [comment]
See also https://github.com/coreos/ignition-dracut/issues/94

**Comment by stbenjam** [comment]
> A lot of overlap here with openshift/enhancements#291

I did read through that but it's very tied to solving the problem by embedding an ignition in an ISO, rather than letting ignition do some things before and after networkmanger is up.

**Comment by stbenjam** [comment]
> See also coreos/ignition-dracut#94

Yea that's a similar problem, but we are using https://coreos.com/os/docs/latest/config-drive.html instead, so I'm not sure this issue helps us

**Comment by cgwalters** [comment]
What I'd argued in one of the epic number of tickets around this is that I think all we need to do for all the major clouds that use the link-local address is bring up networking enough to do that, then we could pull network configuration out of the Ignition itself - then we can do static networking in clouds too (not that most people want this).

That said, if we fixed https://github.com/openshift/machine-config-operator/issues/1690 then you guys could "flatten" the config into `/boot/config.ign` and then you wouldn't need networking in the initramfs at all, right?  I go back and forth on whether or not this is a good idea; it takes away some control from the MCO.

> rather than letting ignition do some things before and after networkmanger is up.

But we'd need to flesh out more what that would look like.  Something like:

```
"initramfs-network": {
    "dhcp": false,
    "args": "ip=192.xxx...."
```

Or just support writing arbitrary things into `/etc/NetworkManager` before it starts, basically a `filesystem` bit that applies to the initramfs?

**Comment by stbenjam** [comment]
> That said, if we fixed openshift/machine-config-operator#1690 then you guys could "flatten" the config into /boot/config.ign and then you wouldn't need networking in the initramfs at all, right? I go back and forth on whether or not this is a good idea; it takes away some control from the MCO.

That addresses the immediate problem for us, but not the general use. It makes *complete* sense to use network data sources for append/merge in ignition, but there's always going to be a circular dependency on environments that have networks that need configuration.

>> rather than letting ignition do some things before and after networkmanger is up.
> But we'd need to flesh out more what that would look like. Something like:
> "initramfs-network": {
>    "dhcp": false,
>    "args": "ip=192.xxx...."
> Or just support writing arbitrary things into /etc/NetworkManager before it > starts, basically a filesystem bit that applies to the initramfs?

Maybe? I don't understand entirely how the initramfs dracut ignition interact -- is the /etc/NetworkManager at that stage in the ramdisk or is that the host's eventual configuration?

**Comment by cgwalters** [comment]
We may need a meeting on this but I really want to walk though how the Live ISO changes the game for you guys.  Hmm....I think we maybe do need to support some flag/mechanism for the Live ISO to disable networking in the initramfs even if Ignition is injected.

> Maybe? I don't understand entirely how the initramfs dracut/ignition stuff interact -- is the /etc/NetworkManager at that stage in the ramdisk or is that the host's eventual configuration?

We use NM in the initrd too (RHCOS will after 8.2).  There's no required linkage between the initramfs configuration and the real root; most people will want them to be the same (see https://github.com/coreos/ignition-dracut/pull/89 ) but like that PR we can add an option to avoid propagating them to the real root.

**Comment by jlebon** [comment]
Related: https://github.com/coreos/ignition/pull/956 (which would also be smart enough to handle live ISO with embedded Ignition configs). It's also an on-ramp for supporting fetching Ignition configs over link-local addresses.

**Comment by jlebon** [comment]
I'm very hesitant to allow Ignition to configure the initramfs itself. I'd prefer we stay away from that for as long as we can because it essentially introduces the same problem for the initrd which Ignition was meant to solve for the real root.

> It makes complete sense to use network data sources for append/merge in ignition, but there's always going to be a circular dependency on environments that have networks that need configuration.

We've been working hard on improving the situation there. :) In addition to the links above, see also https://github.com/coreos/fedora-coreos-tracker/issues/460. I agree with Colin that we should push further on the live ISO embedded config path for bare metal, and investigate the link-local approach for clouds. (I'm also not sure how Ignition can correctly implement the RFE here if e.g. you configure networking in the base config, but then one of the `merge` configs actually wants to override what the config file should have been.)

**Comment by stbenjam** [comment]
> I'm very hesitant to allow Ignition to configure the initramfs itself. I'd prefer we stay away from that for as long as we can because it essentially introduces the same problem for the initrd which Ignition was meant to solve for the real root.

Would it be kernel command line arguments then for this configuration? I wasn't sure if that worked in rhcos, since we tried it when we were trying to get IPv6 working. Can you do VLAN configuration with those arguments?

>> It makes complete sense to use network data sources for append/merge in ignition, but there's always going to be a circular dependency on environments that have networks that need configuration.
> We've been working hard on improving the situation there. :) In addition to the links above, see also coreos/fedora-coreos-tracker#460. I agree with Colin that we should push further on the live ISO embedded config path for bare metal, and investigate the link-local approach for clouds.

The ISO approach for bare metal isn't going to work in the IPI case, and that seems to be ignored in the present designs. We do exist and have customers. We have some initial discussions about how we can use the live ISO for IPI but it's going to be later than we need this problem solved.

I also don't see how the live ISO case fits into OpenShift, because the installer generates a stub ignition with network dependencies and you're going to run into customers with these complex network requirements. I don't see where that use case is being addressed.

>(I'm also not sure how Ignition can correctly implement the RFE here if e.g. you configure networking in the base config, but then one of the merge configs actually wants to override what the config file should have been.)

You could have something like afterburn to be able to do preflight configuration before ignition runs. 

OpenStack config drives have network metadata for similar reasons: https://docs.openstack.org/nova/latest/user/metadata.html.

**Comment by cgwalters** [comment]
> I'm very hesitant to allow Ignition to configure the initramfs itself. I'd prefer we stay away from that for as long as we can because it essentially introduces the same problem for the initrd which Ignition was meant to solve for the real root.

If it's just constrained to networking though, that seems not too bad.

And actually perhaps to start, we could offer exactly one thing - the ability for Ignition to disable initramfs networking.  That would allow using `coreos-installer iso embed` to be fully programmable - the system boots live, the live Ignition can bring up the network however it wants (or not), and run `coreos-installer` however it wants, providing that Ignition config however it wants, etc.

In particular once we have osmet, I suspect a lot of cases could get away with not even bringing up the network at all in the live OS.

**Comment by cgwalters** [comment]
Yeah I am increasingly certain that we missed this important case of "Ignition provided, but does not require networking in the initramfs" in all of the recent work following https://github.com/coreos/fedora-coreos-tracker/issues/443

I took a look at the [recent libvirt testing scripts](https://github.com/coreos/fedora-coreos-config/blob/testing-devel/tests/manual/coreos-network-testing.sh) and it looks like to me theyl use the default bridged network which still has DHCP enabled.  So what happens here is when we're doing a static config in those tests...we still used (and required!) DHCP in the initramfs.

That's not going to work for environments without DHCP at all, or even just ones where we don't want to DHCP on each NIC but not all of them etc.

https://github.com/openshift/enhancements/pull/291 is not complete IMO until we fix this.

Oh right actually, this is exactly what this issue is: https://github.com/coreos/coreos-installer/issues/164#

**Comment by cgwalters** [comment]
OK so yep, going back we have two options:

 - ` coreos-install iso embed --no-initramfs-networking`: https://github.com/coreos/coreos-installer/issues/164#issue-567060872 (only applies to live ISO)
 - Enhance Ignition to have a way for the config to tell us not to use the network at all in the initramfs (this issue)

Or, the much bigger option of:

- Enhance Ignition to have a way to configure initramfs networking

**Comment by dustymabe** [comment]
> Yeah I am increasingly certain that we missed this important case of "Ignition provided, but does not require networking in the initramfs" in all of the recent work following [coreos/fedora-coreos-tracker#443](https://github.com/coreos/fedora-coreos-tracker/issues/443)

A follow-on that we still want to do is https://github.com/coreos/fedora-coreos-tracker/issues/460

> 
> I took a look at the [recent libvirt testing scripts](https://github.com/coreos/fedora-coreos-config/blob/testing-devel/tests/manual/coreos-network-testing.sh) and it looks like to me theyl use the default bridged network which still has DHCP enabled. So what happens here is when we're doing a static config in those tests...we still used (and required!) DHCP in the initramfs.

Can you elaborate what you mean. That test script tests many different configurations including dhcp and static networking in the initramfs combined with static networking in the real root. I do use the default libvirt bridged network, but that's only to cover the dhcp test cases and the machine won't attempt DHCP if you tell it not to. 

> 
> That's not going to work for environments without DHCP at all, or even just ones where we don't want to DHCP on each NIC but not all of them etc.

The EPIC work that I've done specifically focuses on static networking via the installer in the Live ISO. In that case you can easily configure static networking when doing the install.

Follow on work to this is to not bring up networking in the initramfs if not needed at all: https://github.com/coreos/fedora-coreos-tracker/issues/460, but that was not part of the EPIC. 

> 
> [openshift/enhancements#291](https://github.com/openshift/enhancements/pull/291) is not complete IMO until we fix this.
> 
> Oh right actually, this is exactly what this issue is: [coreos/coreos-installer#164](https://github.com/coreos/coreos-installer/issues/164)

I think the better approach is
https://github.com/coreos/fedora-coreos-tracker/issues/460 . For now (RHCOS in 4.6) you can achieve 'no networking in the initramfs' like:

- provide --firstboot-args='' with kargs that disables networking in the initramfs and provide networking config in the real root.

but I guess the real question is why do you need "no networking in the initramfs". I think the real desire is "I only want the networking I want in the initramfs". In that case just configure the networking you want via the installer right?

**Comment by jlebon** [comment]
Heh, was typing up a comment but @dustymabe beat me to it. :)

To summarize: In the general case, I think this is resolved by https://github.com/coreos/fedora-coreos-tracker/issues/460 (which provides a framework for apps to tell the OS that networking is needed) and https://github.com/coreos/ignition/pull/956 + https://github.com/coreos/ignition-dracut/pull/164/ (which makes use of that framework). In the live ISO case/metal case of "complex networking on first boot", one can configure networking as required at install time and use `--copy-network`.

**Comment by dustymabe** [comment]
@jlebon yes. Maybe we should all get together and write up a proper EPIC so the requirements are collected and we can make sure the 460 addresses all needs.

**Comment by cgwalters** [comment]
> provide --firstboot-args='' with kargs that disables networking in the initramfs and provide networking config in the real root.

The user story here is "I have a CoreOS ISO and `coreos-installer iso embed`" and I want to *automate* installs.  `--firstboot-args` is something that runs in the booted OS.  I want to be able to e.g. run a script before or after `coreos-install` without having ever done DHCP at all.

**Comment by cgwalters** [comment]
> To summarize: In the general case, I think this is resolved by coreos/fedora-coreos-tracker#460 (which provides a framework for apps to tell the OS that networking is needed) and #956 + coreos/ignition-dracut#164 (which makes use of that framework).

Right, I added that as a second option in the "more complex" options [above](https://github.com/coreos/ignition/issues/979#issuecomment-632074230).

**Comment by dustymabe** [comment]
> The user story here is "I have a CoreOS ISO and `coreos-installer iso embed`" and I want to _automate_ installs. `--firstboot-args` is something that runs in the booted OS. I want to be able to e.g. run a script before or after `coreos-install` without having ever done DHCP at all.

Yep. We did have https://github.com/coreos/fedora-coreos-config/pull/326 to not require networking for the ISO but we explicitly do bring up networking in the case of an embedded ignition config (as requested during feature discussions) and pointing to future work to take care of that case: https://github.com/coreos/fedora-coreos-config/blob/5615f58487f9fd9f1ebc19e0e6416b53a3e6270f/overlay.d/05core/usr/lib/dracut/modules.d/20live/coreos-liveiso-network-kargs.service#L5-L15

**Comment by cgwalters** [comment]
> and pointing to future work to take care of that case: 

"but the user can override with `rd.neednet=0` now if needed."

No, there's no ergonomic way to override the live ISO kernel arguments in an *automated* fashion.

> We could easily remove the ConditionPathExists=|/config.ign

That's *an* option I guess...I added it to the list above.

**Comment by jlebon** [comment]
>  I want to be able to e.g. run a script before or after coreos-install without having ever done DHCP at all.

Right yeah, this is also in scope for https://github.com/coreos/fedora-coreos-tracker/issues/460. Right now we took the shortcut that embedded config = requires networking. But the end goal is absolutely for Ignition to only ask for networking if required (via the exact same mechanism that is used in clouds/freshly installed metal).

One gap then is if one needs complex networking on first boot of a non-metal platform, and that's what Afterburn is trying to address (by providing a channel for configuring networking, see e.g. https://github.com/coreos/afterburn/pull/404).

Of course, another major gap for much of this is to have osmet supported on RHCOS so that install-time can truly be fully offline. We should chat about our approach on that! (And I see now you recently opened https://github.com/coreos/coreos-assembler/pull/1467 related to this).

**Comment by stbenjam** [comment]
This discussion seems to be going off the rails a bit from our original feature request. I'm not sure why the live CD is part of this particular discussion, we're not using it -- maybe we should have a live discussion about what ignition is missing for us? We need to have something in the 4.6 time frame.

OpenShift creates an ignition that needs networking.  Any ignition has the possibility to include merge/appends that reference network URI's -- what is the answer when a user requires some configuration to make networking work? 

It's impossible today in the baremetal **IPI** platform today to support use cases where the network used to access the MCS requires a VLAN, for example.

**Comment by dustymabe** [comment]
From https://github.com/coreos/ignition/issues/979#issuecomment-632074230

> Another approach is:
>
>   Don't enable networking in the initramfs in the Live ISO at all (but that seems very confusing and breaks the symmetry)

Networking would still be enabled in the other conditions. Just not for the embedded ignition config case.

**Comment by jlebon** [comment]
> maybe we should have a live discussion about what ignition is missing for us?

:+1: Yeah, that'd be great! :)

The reason we're talking about the live ISO is because it will break the dependency loop between Ignition and networking by being able to provide networking configuration upfront at install time.

**Comment by ashcrow** [comment]
> > maybe we should have a live discussion about what ignition is missing for us?
> 
> +1 Yeah, that'd be great! :)

Agreed. The result of this spike should find it's way into a card for execution as well once we've figured out what should be done.

**Comment by cgwalters** [comment]
> This discussion seems to be going off the rails a bit from our original feature request. I'm not sure why the live CD is part of this particular discussion, we're not using it -- maybe we should have a live discussion about what ignition is missing for us? 

Yes, I understand.  If we want to handle your case of using the `-openstack` image (right?) then we can't use option 1.

> OpenShift installer creates an ignition that needs networking. Any ignition has the possibility to include merge/appends that reference network URI's -- what is the answer when a user requires some configuration to make networking work to access those resources?

OK I thought about this more and you're right, we either need to:

1. Make it easy to get a "flattened" real config https://github.com/openshift/machine-config-operator/issues/1690
1. Support configuring the initramfs network via Ignition

Of these...I lean more towards the first, but it's hard.  The reasons (AFAIK) `openshift-install` always generates a pointer configuration is twofold:

 - In e.g. AWS, there are size limitations on user data, but I suspect you don't have that problem
 - The actual Ignition lives inside the MCO codebase and the installer binary just doesn't have it "offline".   Solving that would be...hard and messy; the MCO (most notably the MCS) is just designed to run as part of a cluster.  Really we need something like a bootstrap node to extract that configuration.

Anyways, once we have a flattened config, then as long as we have a way to just *disable* the initramfs networking requirement, we only have to concern ourselves with networking in the real root.

**Comment by stbenjam** [comment]
> > This discussion seems to be going off the rails a bit from our original feature request. I'm not sure why the live CD is part of this particular discussion, we're not using it -- maybe we should have a live discussion about what ignition is missing for us?
> 
> Yes, I understand. If we want to handle your case of using the `-openstack` image (right?) then we can't use option 1.

That's what we use today, I'd expect us to be able to move the IPI install for baremetal to using the live RHCOS image eventually, but not in 4.6.

> > OpenShift installer creates an ignition that needs networking. Any ignition has the possibility to include merge/appends that reference network URI's -- what is the answer when a user requires some configuration to make networking work to access those resources?
> 
> OK I thought about this more and you're right, we either need to:
> 
>     1. Make it easy to get a "flattened" real config [openshift/machine-config-operator#1690](https://github.com/openshift/machine-config-operator/issues/1690)
> 
>     2. Support configuring the initramfs network via Ignition
> 
> 
> Of these...I lean more towards the first, but it's hard. The reasons (AFAIK) `openshift-install` always generates a pointer configuration is twofold:
> 
>     * In e.g. AWS, there are size limitations on user data, but I suspect you don't have that problem

We store the user data in a secret, so we hit the 1MB limitation from Kuberenetes. Ironic's own limitation is 64MB.  1MB is maybe a bit small but it's reasonable for the general case.

>     * The actual Ignition lives inside the MCO codebase and the installer binary just doesn't have it "offline".   Solving that would be...hard and messy; the MCO (most notably the MCS) is just designed to run as part of a cluster.  Really we need something like a bootstrap node to extract that configuration.
> 
> 
> Anyways, once we have a flattened config, then as long as we have a way to just _disable_ the initramfs networking requirement, we only have to concern ourselves with networking in the real root.

This addresses our immediate need.  For the installer, I already wrote up a simple POC that flattens an ignition with terraform: https://github.com/openshift-metal3/terraform-provider-openshift/blob/master/openshift/data_source_inline_ignition_v22.go#L35. We can make that work somehow, we just need a way to access the MCS in the cluster to do the same for workers.

We run into edge cases where users may do things that end up exceeding 1MB for the flattened ignition, which is why the second option of having network support for the initramfs is appealing to me. It also handles the very generalized case for anyone using ignition. Presumably there's some future possibility that ignition ends up in RHEL, or CoreOS is used more generally and we need to think about the generalized need.

**Comment by cgwalters** [comment]
OK some members of the CoreOS team had a live chat about this issue (and we're going to schedule a followup meeting with you guys) but here are some conclusions:

1) There is general agreement that our epics around the Live ISO and static networking are not complete until we have a way to disable the initramfs networking requirement for the live ISO boot
2) We will not add any changes to the OpenStack image or Ignition in general to the epic at this time for 4.6
3) There is general agreement that the "conditional networking" change for Ignition is the most elegant fix (among others it will speed up `cosa run` w/qemu)

I think a corollary to this is we'd like to have a discussion about how hard it really is for you guys to switch to the Live ISO.

**Comment by stbenjam** [comment]
That sounds great, it'd be good to invite @hardys and @juliakreger as well.

**Comment by cgwalters** [comment]
https://github.com/coreos/fedora-coreos-config/pull/415

**Comment by cgwalters** [comment]
Schedule intersections are hard, sent something for 2pm EST next Tuesday, hope that works.

**Comment by kirankt** [comment]
> > This discussion seems to be going off the rails a bit from our original feature request. I'm not sure why the live CD is part of this particular discussion, we're not using it -- maybe we should have a live discussion about what ignition is missing for us?
> 
> Yes, I understand. If we want to handle your case of using the `-openstack` image (right?) then we can't use option 1.
> 
> > OpenShift installer creates an ignition that needs networking. Any ignition has the possibility to include merge/appends that reference network URI's -- what is the answer when a user requires some configuration to make networking work to access those resources?
> 
> OK I thought about this more and you're right, we either need to:
> 
>     1. Make it easy to get a "flattened" real config [openshift/machine-config-operator#1690](https://github.com/openshift/machine-config-operator/issues/1690)

Just wanted to add my 2 cents for this Option 1. We will need this for Day 2 use as well. Once the install is finished, we will still need to get to flattened ignition from Metal3 ecosystem for provisioning additional nodes.

> 
>     2. Support configuring the initramfs network via Ignition
> 
> 
> Of these...I lean more towards the first, but it's hard. The reasons (AFAIK) `openshift-install` always generates a pointer configuration is twofold:
> 
>     * In e.g. AWS, there are size limitations on user data, but I suspect you don't have that problem
> 
>     * The actual Ignition lives inside the MCO codebase and the installer binary just doesn't have it "offline".   Solving that would be...hard and messy; the MCO (most notably the MCS) is just designed to run as part of a cluster.  Really we need something like a bootstrap node to extract that configuration.
> 
> 
> Anyways, once we have a flattened config, then as long as we have a way to just _disable_ the initramfs networking requirement, we only have to concern ourselves with networking in the real root.

**Comment by miabbott** [comment]
Folks from the KNI team and CoreOS team had a meeting today about this issue.  Notes captured here - https://hackmd.io/QqY2xZGVSVqO8L8zBPqM7w

**Comment by cgwalters** [comment]
The TL;DR on this is that for OpenShift <= 4.6 at least we will recommend that anyone who wants to control the initramfs networking for anything except the Live ISO to mount the disk image and edit the kernel cmdline.

We already support networking on the cmdline for PXE cases, so this is fully symmetric.

Future work will focus on work mostly already outlined in https://github.com/openshift/enhancements/pull/291

**Comment by hardys** [comment]
@cgwalters Just wanted to see if we can reach a conclusion on the next steps here, as I'm not clear that we've got a plan to resolve both use-cases in the original issue.

1. Ability to disable networking (or specific interfaces) in the ramdisk - this will be handled via cmdline args (which isn't ideal for IPI as we have one PXE config for all nodes...), and in future this may be improved by openshift/enhancements#291

2. How to retrieve the rendered ignition config from the MCS, when the network it needs is on a non-default VLAN.  It's not clear to me we have any solution for this case yet, and it will potentially be a common requirement for baremetal environments, can you help me understand the next steps here please?  Thanks!

**Comment by cgwalters** [comment]
First I want to be clear - you are specifically requesting this for the OpenStack image right?   In other words this is a request for a targeted "baremetal IPI <= 4.5 quick patch" and not some sort of general change, right?  Because I think the conclusion here is our "general change" is the Live ISO + vmware backchannels already elaborated in the enhancement you linked.

> Ability to disable networking (or specific interfaces) in the ramdisk 

Isn't that covered by https://github.com/coreos/ignition/issues/979#issuecomment-634800899 ?
i.e. "open up and edit the disk" for versions <= 4.5 ?

> How to retrieve the rendered ignition config from the MCS, when the network it needs is on a non-default VLAN. It's not clear to me we have any solution for this case yet, and it will potentially be a common requirement for baremetal environments, can you help me understand the next steps here please? Thanks!

If we definitively conclude we need that, let's discuss in https://github.com/openshift/machine-config-operator/issues/1690

**Comment by ashcrow** [comment]
cc'ing @imcleod for visibility as this is a feature request

**Comment by hardys** [comment]
> First I want to be clear - you are specifically requesting this for the OpenStack image right? In other words this is a request for a targeted "baremetal IPI <= 4.5 quick patch" and not some sort of general change, right? Because I think the conclusion here is our "general change" is the Live ISO + vmware backchannels already elaborated in the enhancement you linked.

We tried the quick-hack for baremetal and it got complex (ref https://github.com/openshift/machine-config-operator/issues/1690 as one issue we encountered) so ideally I think we should try to solve this more generally than some special solution that only works for IPI baremetal?

I don't actually think this necessarily has to be backported or work with older releases, the goal is to provide a supportable interface in some future release.

> > Ability to disable networking (or specific interfaces) in the ramdisk
> 
> Isn't that covered by [#979 (comment)](https://github.com/coreos/ignition/issues/979#issuecomment-634800899) ?
> i.e. "open up and edit the disk" for versions <= 4.5 ?

This approach was rejected by $customer but if it's documented somewhere as a supported approach it may work as an interim solution.  The main issue I can see is you may end up needing multiple custom images if you have pools of nodes with different nic setups.
 
> > How to retrieve the rendered ignition config from the MCS, when the network it needs is on a non-default VLAN. It's not clear to me we have any solution for this case yet, and it will potentially be a common requirement for baremetal environments, can you help me understand the next steps here please? Thanks!
> 
> If we definitively conclude we need that, let's discuss in [openshift/machine-config-operator#1690](https://github.com/openshift/machine-config-operator/issues/1690)

I think this is a fairly common requirement in the case where you want to PXE boot machines, then have the controlplane traffic on a tagged network attached to the same NIC.  I guess this same issue exists for UPI deployments in that case.

If it helps I'm happy to close this and work on an enhancement (or track via https://github.com/openshift/machine-config-operator/issues/1690), this issue was raised I think with the idea that we wanted ignition to grow some "early config" interface, but it seems like it may be preferable to focus on making it easier to inject the combination of the pointer config and the fully rendered one?  

It seems like that would also be required in the ISO embed solution, as you still need to consume the combination of the pointer config (which the user might append config to using the `create ignition-configs` interface of openshift-install) and the rendered config?

**Comment by cgwalters** [comment]
> so ideally I think we should try to solve this more generally than some special solution that only works for IPI baremetal?

We believe that the general solution is the Live ISO.

**Comment by hardys** [comment]
> > so ideally I think we should try to solve this more generally than some special solution that only works for IPI baremetal?
> 
> We believe that the general solution is the Live ISO.

Understood - How do you envisage this working in the PXE boot case?

**Comment by cgwalters** [comment]
> Understood - How do you envisage this working in the PXE boot case?

Via the `ip=` kernel argument - same way this has been done with PXE for a long time, right?

**Comment by cgwalters** [comment]
> It seems like that would also be required in the ISO embed solution, as you still need to consume the combination of the pointer config (which the user might append config to using the create ignition-configs interface of openshift-install) and the rendered config?

No; the Ignition provided to the ISO is just a wrapper around running `coreos-installer` to a target disk plus providing the pointer Ignition, and possibly including any pre-defined networking configuration.

On the second boot, the OpenShift pointer ignition will be used alongside the pre-injected network configuration.

**Comment by hardys** [comment]
>Via the ip= kernel argument - same way this has been done with PXE for a long time, right?

That works when you boot the ramdisk that writes RHCOS to disk, but what about the first boot of RHCOS after that, when it has only the pointer ignition but also requires some network configuration before it can resolve the pointer URL to download the full config?

> On the second boot, the OpenShift pointer ignition will be used alongside the pre-injected network 
configuration.

It sounds like I'm missing something fundamental here - how is that pre-injected network configuration provided to RHCOS for the first boot, so it can resolve the pointer ignition, by customizing the ISO?

That is the crux of the problem we're trying to solve, and I don't think it is dependent on how the RHCOS image gets written to disk, on first boot it still has to download the config via the pointer ignition?

Thanks for the help figuring this out! :)

**Comment by jlebon** [comment]
>> On the second boot, the OpenShift pointer ignition will be used alongside the pre-injected network configuration.
>
> It sounds like I'm missing something fundamental here - how is that pre-injected network configuration provided to RHCOS for the first boot, so it can resolve the pointer ignition, by customizing the deployed RHCOS disk image, or is there some other method?

See https://github.com/coreos/coreos-installer/pull/212 and https://github.com/coreos/fedora-coreos-config/pull/346. IOW, at install-time, you pass `--copy-network` to `coreos-installer install` and that will inject the NM keyfiles into `/boot`. Those keyfiles are then copied into the initramfs at first boot time *before* Ignition needs to fetch anything.

So in essence, your live ISO/PXE Ignition config would include three things:
1. the "base" Ignition config of the target install
2. any NM keyfiles needed for custom networking (note that in the live ISO boot, Ignition won't try to turn on networking in the initrd, so you can use it properly to configure the networking of the live real root directly if that's needed to e.g. fetch the RHCOS metal image or Ignition config)
3. a systemd service that runs `coreos-installer install /dev/foo --ignition base.ign --copy-network ...`

**Comment by hardys** [comment]
> See [coreos/coreos-installer#212](https://github.com/coreos/coreos-installer/pull/212) and [coreos/fedora-coreos-config#346](https://github.com/coreos/fedora-coreos-config/pull/346). IOW, at install-time, you pass `--copy-network` to `coreos-installer install` and that will inject the NM keyfiles into `/boot`. Those keyfiles are then copied into the initramfs at first boot time _before_ Ignition needs to fetch anything.

Thanks @jlebon - this is the detail I was missing :)

I wonder if we could consider making the 15copy-installer-network script copy from a config-drive partition if it exists?  One of the problems when the ironic deploy ramdisk is used (as is currently the case for IPI baremetal) it can write a config-drive partition, but there's no support for modifying the OS disk image itself in the deploy agent that writes to disk.

Long term we can also look at how we might consume the rhcos initrd with coreos-install, but there are some feature gaps and interfaces differences which will need further investigation if we want that to work with the ironic services we currently use.

**Comment by jlebon** [comment]
> I wonder if we could consider making the 15copy-installer-network script copy from a config-drive partition if it exists?

Hmm... maybe. (Though the fact that it's optional makes it similar to https://github.com/coreos/ignition/issues/928.) Let's see if there's a better way first.

> One of the problems when the ironic deploy ramdisk is used (as is currently the case for IPI baremetal) it can write a config-drive partition, but there's no support for modifying the OS disk image itself (so we can't drop files in /boot).

Do you have a link to documentation/code re. what the deployer is capable of? I presume you're not able to run arbitrary code?

> Long term we can also look at how we might consume the rhcos initrd with coreos-install, but there are some feature gaps and interfaces differences which will need further investigation if we want that to work with the ironic services we currently use.

Yes, agreed! :+1:

**Comment by dustymabe** [comment]
> Long term we can also look at how we might consume the rhcos initrd with coreos-install, but there are some feature gaps and interfaces differences which will need further investigation if we want that to work with the ironic services we currently use.

Long term I think there are a few options with varying levels of difficulty so we might progressively go through them:

1. teach the deployer how to mount /boot from the disk that was recently imaged and modify contents
    - This is the easiest to implement, but not ideal as the deployer is now responsible for knowing internal details of how we are moving things around in RHCOS/FCOS land.
2. include the `coreos-installer` binary in the deployer and have the deployer call `coreos-installer` with specified options to image the disk
3. have the deployer actually be the RHCOS/FCOS live PXE/ISO environment in the future.
    - an ignition config can be provided to check in with services and such (see https://dustymabe.com/2020/04/04/automating-a-custom-install-of-fedora-coreos/)

**Comment by hardys** [comment]
> > One of the problems when the ironic deploy ramdisk is used (as is currently the case for IPI baremetal) it can write a config-drive partition, but there's no support for modifying the OS disk image itself (so we can't drop files in /boot).
> 
> Do you have a link to documentation/code re. what the deployer is capable of? I presume you're not able to run arbitrary code?

There is an overview here https://docs.openstack.org/ironic/latest/user/index.html#understanding-bare-metal-deployment - the agent itself is https://github.com/openstack/ironic-python-agent which is contained in a ramdisk we use to write the OCP disk image (similar to the OCP ramdisk with coreos-install).

Basically there are no interfaces for running arbitrary code, or injecting arbitrary files, the baremetal box is treated like a cloud VM - you provide an OS disk image, and optionally some user-data/meta-data (which in the metal3 case happens via a config-drive partition)

OpenStack has support for a network_data.json ref https://docs.openstack.org/nova/latest/user/metadata.html#openstack-format-metadata but I don't think the ignition openstack backend does anything with that atm?

**Comment by hardys** [comment]
> > Long term we can also look at how we might consume the rhcos initrd with coreos-install, but there are some feature gaps and interfaces differences which will need further investigation if we want that to work with the ironic services we currently use.
> 
> Long term I think there are a few options with varying levels of difficulty so we might progressively go through them:
> 
>     1. teach the deployer how to mount /boot from the disk that was recently imaged and modify contents
>        
>        * This is the easiest to implement, but not ideal as the deployer is now responsible for knowing internal details of how we are moving things around in RHCOS/FCOS land.

Yeah I suspect that won't be acceptable to upstream Ironic folks as a change to ironic-python-agent, since it's designed to be completely OS agnostic /cc @juliakreger @dtantsur
> 
>     2. include the `coreos-installer` binary in the deployer and have the deployer call `coreos-installer` with specified options to image the disk
> 
>     3. have the deployer actually be the RHCOS/FCOS live PXE/ISO environment in the future.
>        
>        * an ignition config can be provided to check in with services and such (see https://dustymabe.com/2020/04/04/automating-a-custom-install-of-fedora-coreos/)

I suspect this is where we want to get to, the main challenge is right now we depend on some features of ironic-python-agent (in particular introspection), so the RHCOS/FCOS ramdisk isn't a drop-in replacement.  

Switching to an RHCOS/FCOS based solution would either require reimplementing some of the IPA features (and interfaces to interact with ironic, but that could be a shim injected like in your blog post I guess), or maybe have a way to run IPA on top of the RHCOS/FCOS image (which doesn't solve the inject-files-to-boot problem but it has been done before ref https://docs.openstack.org/ironic-python-agent/queens/install/index.html#coreos)

**Comment by dtantsur** [comment]
> the deployer is now responsible for knowing internal details of how we are moving things around in RHCOS/FCOS land.

Yeah, it would practically require a downstream addition to the ramdisk. I'd like to note that ironic-python-agent is designed to be pretty flexible, so it's not an entirely bad thing either. We're working on a thing called "deploy steps" specifically to allow site-specific customizations.

> it has been done before

I think CoreOS was so different at the time that it won't help us (and we did some ugly things like chroot in tarball contents).

**Comment by jlebon** [comment]
> the baremetal box is treated like a cloud VM - you provide an OS disk image, and optionally some user-data/meta-data (which in the metal3 case happens via a config-drive partition)

OK right, that makes a lot of sense now.

> I suspect this is where we want to get to, the main challenge is right now we depend on some features of ironic-python-agent (in particular introspection), so the RHCOS/FCOS ramdisk isn't a drop-in replacement.

So to recap, I think we're agreed on where we want to go long-term. And the question now is how to get this working in the short-term with the existing constraints of the deployer. @hardys Do I understand correctly that the "edit image" hack should work for now?

If we can do the flanned Ignition config thing, this could also be solved by https://github.com/coreos/fedora-coreos-config/pull/426.

**Comment by hardys** [comment]
> So to recap, I think we're agreed on where we want to go long-term. And the question now is how to get this working in the short-term with the existing constraints of the deployer. @hardys Do I understand correctly that the "edit image" hack should work for now?
> 
> If we can do the flattened Ignition config thing, this could also be solved by [coreos/fedora-coreos-config#426](https://github.com/coreos/fedora-coreos-config/pull/426).

Sorry for the delay responding @jlebon yes I think that's accurate, so to summarize:

* Short term workaround is to edit the RHCOS image (we'll need some docs/kbase downstream for this)
* Medium term look to use https://github.com/coreos/fedora-coreos-config/pull/426 combined with the flattened config discussed in https://github.com/openshift/machine-config-operator/issues/1690 (hopefully we can get this into plans for OKD 4.7?)
* Long term, look to align the IPI deploy ramdisk with the coreos-install based solution (as mentioned there are significant issues to resolve here so we'll have to figure those out which may take some time/prioritization)

@stbenjam @kirankt does that sound reasonable?

**Comment by dtantsur** [comment]
> Short term workaround is to edit the RHCOS image (we'll need some docs/kbase downstream for this)

In recent ironic (~ a week ago) you can drop a python plugin to IPA that will be called automatically during deployment to do.. pretty much whatever you want. We can have a downstream plugin that does $stuff for us as a short-term workaround.

**Comment by hardys** [comment]
> > Short term workaround is to edit the RHCOS image (we'll need some docs/kbase downstream for this)
> 
> In recent ironic (~ a week ago) you can drop a python plugin to IPA that will be called automatically during deployment to do.. pretty much whatever you want. We can have a downstream plugin that does $stuff for us as a short-term workaround.

@dtantsur sounds interesting - can you link any docs/examples?  Could we modify /boot (sed the grub.cfg and/or drop some file in there) after IPA writes the OS image?  What would the interface from ironic look like?

**Comment by dtantsur** [comment]
> sounds interesting - can you link any docs/examples?

It's a brand new features, I'm going to work on examples this/next week.

> Could we modify /boot (sed the grub.cfg and/or drop some file in there) after IPA writes the OS image?

Yes! But do you really need it to happen afterwards or do you need to modify `/etc/defaults/grub` before grub installation? I think the latter is a bit more robust.

> What would the interface from ironic look like?

There are options. If we create an openshift-specific plugin, it may have no interface at all, just silently do what we need. If you could remind me what exactly we need to do (I'm a bit lost in the history of this issue), I can say specifically.

**Comment by jlebon** [comment]
>> Could we modify /boot (sed the grub.cfg and/or drop some file in there) after IPA writes the OS image?
>
> Yes! But do you really need it to happen afterwards or do you need to modify `/etc/defaults/grub` before grub installation? I think the latter is a bit more robust.

Note FCOS and RHCOS don't use `grub2-mkconfig`, but pure BLS: https://www.freedesktop.org/wiki/Specifications/BootLoaderSpec/. So the canonical kargs live in the BLS configs at `$(boot)/loader/entries/` and that's what you'd typically modify after install. But here we're mostly talking about mounting boot in order to drop network configs in there.

All these are things that `coreos-installer` has native support for (via e.g. `--copy-network`, `--append-karg`, `--delete-karg`), so once we switch over to that for the install, it'll be easier/safer to do these modifications.

**Comment by dtantsur** [comment]
> All these are things that coreos-installer has native support for (via e.g. --copy-network, --append-karg, --delete-karg), so once we switch over to that for the install, it'll be easier/safer to do these modifications.

I'm not sure it's going to happen, at least not in all scenarios.

**Comment by hardys** [comment]
> All these are things that `coreos-installer` has native support for (via e.g. `--copy-network`, `--append-karg`, `--delete-karg`), so once we switch over to that for the install, it'll be easier/safer to do these modifications.

@jlebon can you link any docs on how this works in an automated environment?  E.g if you pxe boot the installer ramdisk, how does this data get provided to coreos-installer for a non-interactive deploy?

As @dtantsur says there is some uncertainty/challenge in making this change (hence I mention it as a long term plan above), so we'll need to investigate further.

We'll need to document a workaround in the meantime - does anything exist (upstream or downstream) that describes the process of using virt-customize to drop additional network configuration in /boot? (I know the virt-customize commands, but not certain on the network file format expected, and we need something downstream to share with $customer).

**Comment by jlebon** [comment]
> @jlebon can you link any docs on how this works in an automated environment? E.g if you pxe boot the installer ramdisk, how does this data get provided to coreos-installer for a non-interactive deploy?

We don't have good docs yet for that right now, but in an automated environment, you would script a systemd unit which calls `coreos-installer install --copy-network --append-karg foo ...`.

If you're rolling your own code, you can of course do the same thing `coreos-installer` does when handling these args. For `--copy-network`, you would copy the NM keyfiles into `$bootmnt/coreos-firstboot-network` (see the code that handles this on firstboot here: https://github.com/coreos/fedora-coreos-config/blob/87f5c512ad60e4f37c2b10434dda2e59e40dd97a/overlay.d/05core/usr/lib/dracut/modules.d/15coreos-network/coreos-copy-firstboot-network.sh). For kargs, you can modify the `options` line in the BLS configs in `$bootmnt/loader/entries`.

**Comment by hardys** [comment]
> Since [coreos/fedora-coreos-config#426](https://github.com/coreos/fedora-coreos-config/pull/426) merged (fixes [coreos/fedora-coreos-tracker#443](https://github.com/coreos/fedora-coreos-tracker/issues/443)) it feels like the question of how to **"Support disabling initramfs networking via Ignition"** is answered.
> 
>     * Q: How do I disable initramfs networking via Ignition?
> 
>     * A: You provide an Ignition config that doesn't have any remote references.
>       
>       * This implies it doesn't require networking and no networking will be brought up in the initramfs.
> 
> 
> The caveat to this is that there are some platforms that only get their Ignition configuration by using the network, so it's not really possible to disable it on those platforms, nor would you want to.

I pushed https://github.com/openshift/enhancements/pull/467 so we can discuss some ideas around the "flattened ignition" solution, which (where appropriate, and in particular for IPI baremetal) could enable a MCO managed Ignition config that doesn't have any remote references

> Shall we close this issue out?

It seems the the conclusion of the discussion so far is that we won't fix this by adding some new ignition interface for early network config, so that seems reasonable to me.

**Comment by dustymabe** [comment]
Thanks @hardys - Will follow along the discussion in https://github.com/openshift/enhancements/pull/467

**Comment by dustymabe** [comment]
FTR the work for #433 was first introduced in testing stream release `32.20200715.2.2` and stable stream release `32.20200715.3.0`.
